services/logs/filter.go

```diff
@@ -0,0 +1,216 @@
+	"slices"
+)
+
+type FilterCriteria struct {
+	Addresses []common.Address
+	Topics    [][]common.Hash
+}
+
+// todo think about creating an interface for all the filters
+// Filter interface { Match() (chan *gethTypes.Log, error) }
+
+// RangeFilter matches all the indexed logs within the range defined as
+// start and end block height. The start must be strictly smaller than end value.
+type RangeFilter struct {
+	start, end *big.Int
+	criteria   FilterCriteria
+	receipts   storage.ReceiptIndexer
+}
+
+func NewRangeFilter(
+	start, end big.Int,
+	criteria FilterCriteria,
+	receipts storage.ReceiptIndexer,
+) *RangeFilter {
+	return &RangeFilter{
+		start:    &start,
+		end:      &end,
+		criteria: criteria,
+		receipts: receipts,
+	}
+}
+
+func (r *RangeFilter) Match() ([]*gethTypes.Log, error) {
+	if r.start.Cmp(r.end) != -1 {
+		return nil, fmt.Errorf("invalid start and end block height, start must be smaller than end value")
+	}
+
+	blooms, heights, err := r.receipts.BloomsForBlockRange(r.start, r.end)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(blooms) != len(heights) {
+		return nil, fmt.Errorf("bloom values don't match height values") // this should never happen
+	}
+
+	logs := make([]*gethTypes.Log, 0)
+	for i, bloom := range blooms {
+		if !bloomMatch(bloom, r.criteria) {
+			continue
+		}
+
+		// todo do this concurrently
+		receipt, err := r.receipts.GetByBlockHeight(heights[i])
+		if err != nil {
+			return nil, err
+		}
+
+		for _, log := range receipt.Logs {
+			if exactMatch(log, r.criteria) {
+				logs = append(logs, log)
+			}
+		}
+	}
+
+	return logs, nil
+}
+
+// IDFilter matches all logs against the criteria found in a single block identified
+// by the provided block ID.
+type IDFilter struct {
+	id       common.Hash
+	criteria FilterCriteria
+	blocks   storage.BlockIndexer
+	receipts storage.ReceiptIndexer
+}
+
+func NewIDFilter(
+	id common.Hash,
+	criteria FilterCriteria,
+	blocks storage.BlockIndexer,
+	receipts storage.ReceiptIndexer,
+) *IDFilter {
+	return &IDFilter{
+		id:       id,
+		criteria: criteria,
+		blocks:   blocks,
+		receipts: receipts,
+	}
+}
+
+func (i *IDFilter) Match() ([]*gethTypes.Log, error) {
+	blk, err := i.blocks.GetByID(i.id)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := i.receipts.GetByBlockHeight(big.NewInt(int64(blk.Height)))
+	if err != nil {
+		return nil, err
+	}
+
+	logs := make([]*gethTypes.Log, 0)
+	for _, log := range receipt.Logs {
+		if exactMatch(log, i.criteria) {
+			logs = append(logs, log)
+		}
+	}
+
+	return logs, nil
+}
+
+// StreamFilter matches all the logs against the criteria from the receipt channel.
+type StreamFilter struct {
+	criteria      FilterCriteria
+	receiptStream chan *gethTypes.Receipt
+}
+
+func NewStreamFilter(criteria FilterCriteria, receipts chan *gethTypes.Receipt) *StreamFilter {
+	return &StreamFilter{
+		criteria:      criteria,
+		receiptStream: receipts,
+	}
+}
+
+func (s *StreamFilter) Match() (chan *gethTypes.Log, error) {
```

Comment:
best to only return the channel receiver, otherwise it could be written to or closed

```suggestion
func (s *StreamFilter) Match() (<-chan *gethTypes.Log, error) {
```

Reply:
good point